frb_rust/src/ffi/web.rs

```diff
@@ -0,0 +1,207 @@
+        self.clone().into_dart()
+    }
+}
+
+#[derive(Clone)]
+pub struct Isolate {
+    port: MessagePort,
+}
+
+impl Isolate {
+    pub fn new(port: MessagePort) -> Self {
+        Self { port }
+    }
+    pub fn post(&self, msg: impl IntoDart) -> bool {
+        self.port
+            .post_message(&msg.into_dart())
+            .map_err(|err| {
+                crate::console_error!("post: {:?}", err);
+                err
+            })
+            .is_ok()
+    }
+}
+
+#[wasm_bindgen]
+extern "C" {
+    #[wasm_bindgen(js_namespace = console)]
+    pub fn log(msg: &str);
+    #[wasm_bindgen(js_namespace = console)]
+    pub fn error(msg: &str);
+    // fn eval(script: &str);
+}
+
+type RawClosure<T> = Box<dyn FnOnce(&[T]) + Send + 'static>;
+
+pub struct TransferClosure<T> {
+    transfer: Vec<T>,
+    closure: RawClosure<T>,
+}
+
+pub struct TransferClosurePayload<T> {
+    func: RawClosure<T>,
+}
+
+impl TransferClosure<JsValue> {
+    pub fn new(transfer: Vec<JsValue>, closure: impl FnOnce(&[JsValue]) + Send + 'static) -> Self {
+        let closure = Box::new(closure);
+        Self { transfer, closure }
+    }
+    /// Posts a <code>[*mut [TransferClosurePayload], ...[JsValue]]</code> message to this worker.
+    ///
+    /// The worker's `onmessage` should run the corresponding [`receive_transfer_closure`]
+    /// to receive the message.
+    pub fn apply(self, worker: &Worker) -> Result<(), JsValue> {
+        let transfer = Array::from_iter(self.transfer);
+        // The worker is responsible for cleaning up the leak here.
+        let payload = Box::into_raw(Box::new(TransferClosurePayload { func: self.closure }));
+        let data = Array::from(&transfer);
+        data.unshift(&JsValue::from(payload as i32));
+        worker
+            .post_message_with_transfer(&data, &transfer)
+            .map_err(|err| {
+                // post message failed, ownership remains with us.
+                drop(unsafe { Box::from_raw(payload) });
+                err
+            })
+    }
+}
+
+pub struct ZeroCopyBuffer<T>(pub T);
+
+impl<T> ZeroCopyBuffer<Vec<T>> {
+    #[inline]
+    pub fn as_slice(&self) -> &[T] {
+        self.0.as_slice()
+    }
+}
+
+/// Internal implementations for transferables on WASM platforms.
+pub trait FromDart {
+    fn from_dart(value: &JsValue) -> Self;
+}
+
+impl<T: JsCast + Clone> FromDart for Option<T> {
+    #[inline]
+    fn from_dart(value: &JsValue) -> Self {
+        value.dyn_ref().cloned()
+    }
+}
+
+macro_rules! delegate_from_dart {
+    ($($ty:ty)*) => {$(
+        impl FromDart for $ty {
+            #[inline]
+            fn from_dart(value: &JsValue) -> Self {
+                value.unchecked_ref::<Self>().clone()
+            }
+        }
+    )*};
+}
+
+delegate_from_dart! {
+    MessagePort ArrayBuffer
+}
+
+/// ## Safety
+/// This function reclaims a raw pointer created by [`TransferClosure`], and therefore
+/// should **only** be used in conjunction with it.
```

Comment:
Thus maybe we can try to avoid the `pub fn` (but use `fn` instead), o/w people may wrongly use it?

Reply:
wasm_bindgen cannot generate code for non-pub functions yet, so sadly no.